### PR TITLE
Overlay Driver

### PR DESCRIPTION
**Pull request by mrjana**
This PR implements the overlay driver and it's support functions to enable native multi host connectivity for docker containers. The driver is very much alpha quality and should be used for demo and testing purposes only.

### CODE REVIEW

File: cmd/ovrouter/ovrouter.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	ep.mac = mac
+	return nil
+}
+
+func (ep *endpoint) InterfaceNames() []driverapi.InterfaceNameInfo {
+	return []driverapi.InterfaceNameInfo{ep}
+
+}
+
+func (ep *endpoint) SetNames(srcName, dstPrefix string) error {
```
Comment (mavenugo) [inline cmd/ovrouter/ovrouter.go]: SetName ?

File: drivers/overlay/ov_utils.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		LinkAttrs: netlink.LinkAttrs{Name: name},
+		VxlanId:   int(vni),
+		Learning:  true,
+		Proxy:     true,
+		L3miss:    true,
+		L2miss:    true,
+	}
+
+	if err := netlink.LinkAdd(vxlan); err != nil {
+		return "", fmt.Errorf("error creating vxlan interface: %v", err)
```
Comment (mavenugo) [inline drivers/overlay/ov_utils.go]: Maybe we shouldn't throw error if vxlan interface already exists ?
Atleast I got this error once  during my testing : `error : network sandbox initialization failed: error creating vxlan interface: file exists`

File: drivers/overlay/joinleave.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+	sbox := n.sandbox()
+
+	name1, name2, err := createVethPair()
+	if err != nil {
+		return err
+	}
+
+	if err := sbox.AddInterface(name1, "veth",
+		sbox.InterfaceOptions().Master("bridge1")); err != nil {
```
Comment (mavenugo) [inline drivers/overlay/joinleave.go]: am seeing a panic when attaching a sandbox to the only endpoint published on a overlay network. Need to dig further.

```
runtime error: invalid memory address or nil pointer dereference
goroutine 116 [running]:
net/http.func·011()
	/usr/local/go/src/net/http/server.go:1130 +0xbb
github.com/docker/libnetwork/drivers/overlay.(*driver).Join(0xc208052bd0, 0xc2080dd840, 0x40, 0xc20819bb80, 0x40, 0xc2081ee860, 0x19, 0x7fdb481ba0f8, 0xc208052d80, 0x0, ...)
	/home/vagrant/workspace/src/github.com/docker/libnetwork/drivers/overlay/joinleave.go:42 +0x5e7
github.com/docker/libnetwork.(*endpoint).Join(0xc208052d80, 0xc2081f7080, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/workspace/src/github.com/docker/libnetwork/endpoint.go:402 +0x7a3
```

File: cmd/ovrouter/ovrouter.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	opt := make(map[string]interface{})
+	if len(os.Args) > 1 {
+		opt[netlabel.OverlayBindInterface] = os.Args[1]
+	}
+	if len(os.Args) > 2 {
+		opt[netlabel.OverlayNeighborIP] = os.Args[2]
+	}
+	r.d.Config(opt)
```
Comment (mavenugo) [inline cmd/ovrouter/ovrouter.go]: This is an useful utility. But, given that the datastore is used now for vxlan and ipam, can you add more configuration for the datastore cfg options to make this useful ?

File: drivers/overlay/overlay.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+			d.store, err = datastore.NewDataStore(cfg)
+			if err != nil {
+				err = fmt.Errorf("failed to initialize data store: %v", err)
+				return
+			}
+		}
+
+		d.vxlanIdm, err = idm.New(d.store, "vxlan-id", vxlanIDStart, vxlanIDEnd)
+		if err != nil {
+			err = fmt.Errorf("failed to initialize vxlan id manager: %v", err)
```
Comment (mavenugo) [inline drivers/overlay/overlay.go]: After a restart (using dnet): 
`Error starting dnetDaemon : failed to initialize vxlan id manager: invalid character '\x00' looking for beginning of value`

**Comment by sridhardevarapalli** [comment]
Is there a design document that has more details on the design of the overlay driver? The link below is empty - https://github.com/docker/libnetwork/blob/master/docs/overlay.md

thanks

**Comment by mavenugo** [comment]
@mrjana LGTM.
Just 1 minor issue after a restart. It might be an issue in `idm` and not necessarily part of this PR.

:+1:

**Comment by mavenugo** [comment]
@sridhardevarapalli we will get it done soon.

**Pull request merged**